### ISSUE TITLE
Auto-rewrite URLs with route prefix defined

### ISSUE THREAD

**Issue by lannonbr** [rooseveltframework/roosevelt#1101]
With the `routePrefix` tag, you can add a URL prefix to any HTTP route set up in a controller. Although, when setting up links to statics or other routes hardcoded in HTML, it is not automatically redirected / rewritten. From the docs, this can be done programatically:

> Example: An image located at /images/teddy.jpg can be resolved in a prefix-agnostic way via `${app.get('routePrefix')}/images/teddy.jpg`.

That said, doing this across all `<a>`, `<link>`, and other tags in the templates can get cluttered. As a possible solution, a middleware function (that can be configured by a roosevelt config setting) could be enabled to automatically rewrite requests on the server to append the route prefix if it hasn't already been added.

So for example with a prefix of `/prefix`:

- If `/prefix/endpoint` is requested, it already has the prefix so it will route properly,
- If `/css/style.css` is requested, roosevelt will append the prefix and it will request `/prefix/css/style.css` instead.

So with this, we would need to add a middleware function to do these rewrites that could be enabled by default but can be modified with a possible `routePrefixRewrite` config field.

My team is willing to do the work to implement this.

**Comment by alallier** [comment]
I can see the desire to not want this on every route. @lannonbr and I talked through all the potential options last week.

Basically what I think we need to do here to cover your case Eric is to support a controller glob "blacklist". We can define controllers that shouldn't auto-redirect. Then when specifying in the client those URLs will require an explicit URL prefix to be concatenated. Personally I don't know how you don't do that as a server config item. There isn't any logic for the client slide HTML to know if it should honor a auto redirecting prefix or not.

Proposed configs and options below:

## Option 1

(Closest to what is currently in Roosevelt)

In this example everything auto redirects except for the speficied controllers/file/paths in the blacklist.

Note: to prevent auto redirect on all routes we could either specify `*` for a glob pattern in the blacklist or provide a higher level config key like `autoRedirect: false`

```js
"prefix": "/foo",
"routeAutoPrefixRedirectBlacklist": [
  "homepage.js",
  "subfolder"
]
```

## Option 2
We could extend upon the work in option 1 and also provide a prefix blacklist as well if you wanted. In this example ever route honors the prefix except the defined items in `routePrefixBlacklist` and every route auto directs on that prefix except the items in `routeAutoPrefixRedirectBlacklist`

Same as above a greedy glob pattern like `*` could be used to block all or a separate config key would be needed

```js
routePrefix: '/foo',
routePrefixBlacklist: [
  './mvc/controllers/fetchAdapters.js',
  './mvc/controllers/robots.txt.js',
  './mvc/controllers/404.js',
  './mvc/controllers/metrics/*'
],
routePrefixAutoRedirectBlacklist: [
 'homepage.js',
 'subfolder'
]
```

## Option 3

We didn't realize you walked back the more dynamic router config in #875 until exploring the potential options

Although more complex this configuration would allow for the _most dynamic_ router config in Roosevelt

```js
"routers": {
  autoRedirect: false,
  "controllers": [
    {
      "prefix": "/foo",
      "files": [
        "homepage.js",
        "subfolder": {
          autoRedirect: false
        }
      ]
    },
    {
      "prefix": "/baz",
      autoRedirect: false,
      "files": [
        "homepage.js": {
          autoRedirect: true
        },
        "subfolder"
      ]
    },
    {
      "prefix": "/bar",
      "files": [
        "bar/controller1.js",
        "bar/controller2.js"
      ]
    }
  ],
  "statics": [
    {
      "prefix": "/public",
      "files": [
        "css", 
        "images", 
        "js"
      ]
    }
  ]
}
```

**Comment by kethinov** [comment]
Talked with @alallier on a call and unpacked what I meant about the need for this to be controlled by the templates, not the server code. That's where the exceptions should be.

To that end, we discussed an alternative proposal that would preserve the basic idea of adding a middleware to Roosevelt implement a route prefix by default and the method of client-side controlled opt-out we discussed would be examining the Express route stack to see if the route being requested by the frontend is already a valid route. If it is, we do not add the prefix. If it isn't and would otherwise 404, then we add the prefix.

This would also allow us to keep the configuration very simple (no wall of text like the 3 options proposed above).

@lannonbr let us know what you think of this idea.
